core/src/main/java/org/apache/calcite/jdbc/CalciteConnectionImpl.java

```diff
@@ -420,6 +421,7 @@ public CalciteServerStatement getStatement(Meta.StatementHandle h)
           .put(Variable.TIME_ZONE.camelName, timeZone)
           .put(Variable.STDIN.camelName, streamHolder.get()[0])
           .put(Variable.STDOUT.camelName, streamHolder.get()[1])
+          .put("#DELTA#", new LinkedList<>()) // TODO rql cleaner way?
```

Comment:
I guess this can be done inside EnumerableDeltaTableScan#implement, no?

Reply:
I think when we arrive there it is "too late": at that point, we can only read from the DataContext via 'get' method, but we cannot put new parameters in there.

Reply:
What if we have recursive (recursive) query?
Then a single DELTA would be not enough.

By the way: please never ever use `LinkedList`.
Here's what the author says: https://twitter.com/joshbloch/status/583813919019573248

Reply:
Another alternative would be to add a new Table during EnumerableDeltaTableScan#implement using DataContext#getRootSchema. This can solve also the problem of having multiple levels of recursion since each different scan can establish a different table name. How about something in these lines?

Reply:
Is it really required to add a table?

For instance, EnumerableJoin manages to operate yet it doesn't require adding a table for that.

Reply:
In this case data have to shared/transferred between different operators (i.e., Union + Scan) which do not necessarily have parent-child relationships. Due to this I would say that it is necessary to introduce "something" that can be shared between operators and I think that a Table would be a good abstraction but maybe I am missing something.

Reply:
@zabetak , I wonder if we should add a new `RelNode` subtype for "recursive with delta" table instead.

Adding tables to the schema for a sake of a single query on the fly does not seem right.

Reply:
The current solution is a naive one, which, as you say, does not support multiple recursive declarations.
In order to address this issue, and following the current approach, we could create each deltaScan with a certain parameter to provide a unique identifier (what about a ModifiableTable?), and then this parameter will have to be transferred also in the moment of the recursiveUnion creation, so that it can be aware of "where to put" the result data (so that it can be picked by the deltaScan in the next iteration).

Reply:
We don't really need an additional table. In Postgres' implementation, RecursiveUnion and DeltaTableScan both have a pointer pointing to each other. DeltaTableScan stores the intermediate data store that RecursiveUnion can access.

Reply:
Committed new version:
- deltaTableScan and recursiveUnion share an identifier (the shared delta table) to carry out the data transfer, this allows to have multiple (nested) recursive unions, unit tests provided showing this scenario.
- implemented (although not used yet) new parameter maxDepth in deltaTableScan (might be useful for graph traversal)

Still TODO: 
- Data transfer between deltaTableScan and recursiveUnion is still done via a parameter in the DataContext, this has to be modified. I'll explorer the approach proposed by @zabetak of using the SchemaPlus via DataContext#getRootSchema

Reply:
Committed a new version that address the TODO mentioned in the comment above: instead of using the DataContext, now the data store / transfer between recursiveUnion and deltaTableScan is done via a transient table (a simple ModifiableTable) that gets added to the schema.

Reply:
Thanks @rubenada! I like the new approach. Let's see what @vlsi , and @hsyuan have to say about this.